contracts/hooks/FullRange.sol

```diff
@@ -0,0 +1,469 @@
+
+        address poolToken;
+        assembly {
+            poolToken := create2(0, add(bytecode, 32), mload(bytecode), salt)
+        }
+
+        poolToERC20[key.toId()] = poolToken;
+
+        return FullRange.beforeInitialize.selector;
+    }
+
+    function _rebalance(IPoolManager.PoolKey calldata key) internal {
+        console.log("we're rebalancing");
+        // retrieve all liquidity -- i think i should already have the liquidity kept track of
+        // uint128 hookLiquidity = poolManager.getLiquidity(key.toId(), address(this), MIN_TICK, MAX_TICK);
+
+        // position.liquidity = hookLiquidity;
+
+        HookPosition storage position = poolToHookPosition[key.toId()];
+
+        IPoolManager.ModifyPositionParams memory params = IPoolManager.ModifyPositionParams({
+            tickLower: MIN_TICK,
+            tickUpper: MAX_TICK,
+            liquidityDelta: -int256(int128(position.liquidity))
+        });
+
+        console.log("the entire hook liquidity");
+        console.log(position.liquidity);
+
+        BalanceDelta balanceDelta = hookModifyPosition(key, params);
+
+        (uint160 sqrtPriceX96,,,,,) = poolManager.getSlot0(key.toId());
+
+        uint128 liquidity = LiquidityAmounts.getLiquidityForAmounts(
+            sqrtPriceX96,
+            TickMath.getSqrtRatioAtTick(MIN_TICK),
+            TickMath.getSqrtRatioAtTick(MAX_TICK),
+            uint256(uint128(-balanceDelta.amount0())),
+            uint256(uint128(-balanceDelta.amount1()))
+        );
+
+        params = IPoolManager.ModifyPositionParams({
+            tickLower: MIN_TICK,
+            tickUpper: MAX_TICK,
+            liquidityDelta: int256(int128(liquidity))
+        });
+
+        // reinvest everything
+        hookModifyPosition(key, params);
+
+        // update position
+        Position.Info memory posInfo = poolManager.getPosition(key.toId(), address(this), MIN_TICK, MAX_TICK);
+
+        position.feeGrowthInside0LastX128 = posInfo.feeGrowthInside0LastX128;
+        position.feeGrowthInside1LastX128 = posInfo.feeGrowthInside1LastX128;
+        position.tokensOwed0 = 0;
+        position.tokensOwed1 = 0;
+    }
+
+    function beforeModifyPosition(
+        address sender,
+        IPoolManager.PoolKey calldata key,
+        IPoolManager.ModifyPositionParams calldata params
+    ) external override returns (bytes4) {
+        // check msg.sender
+        require(sender == address(this), "sender must be hook");
+
+        // check full range
+        // require(
+        //     params.tickLower == MIN_TICK && params.tickUpper == MAX_TICK, "Tick range out of range or not full range"
+        // );
+
+        if (block.number > blockNumber) {
+            console.log("block number increased");
+            blockNumber = block.number;
+            HookPosition storage position = poolToHookPosition[key.toId()];
+
+            if (position.tokensOwed1 > 0 || position.tokensOwed0 > 0) {
+                _rebalance(key);
```

Comment:
you can DRY up the code by putting _all_ the checks here in `_rebalance()` since they just repeat in `beforeSwap` 

so this would just be 
```
require(sender == address(this), "sender must be hook");
_rebalance(key);
```